### PR TITLE
Refactor Worlds to contain a mutable list of Layers.

### LINKED ISSUE #14: Refactor worlds and levels
Background layer -- should be adapted to be an image
Midground layer -- no texture blending, platformer blocks
Entity layer -- normal entities drawn here
Foreground layer -- sprites for ambient effects (ignore for now)

Screen should scroll both horizontally and vertically with the player.

### PR DESCRIPTION
**Pull request by viking-sudo-rm**
1. Worlds now contain a list of layers which are configurable via config.yml.
2. Layer interface standardizes implementation for `BackgroundLayer`s, `TileLayer`s, and `EntityLayer`s. `Level` has been renamed `TileLayer`.
3. Reorganized interface hierarchy in `snorri.world` package. This affects `Savable`, `Loadable`, `Renderable`, etc.
4. Removed old spaghetti code for deprecated features (creating walls, 2D pathfinding).
5. Introduced new static factory pattern for creating worlds and some layers (did not take the time to adapt old stuff).
6. Changed `LevelEditor` to have a currently selected layer for editing. Currently, this layer is set to the standard tile layer, but we can introduce functionality to switch it to edit foreground layers in the future.
5. Other related cleaning up and refactoring.

Through manual testing, I have confirmed that old-formatted levels can be loaded in the `LevelEditor`. Additionally, new levels can be created, and levels that have been loaded can be saved correctly.

Fixes #14.

### CODE REVIEW

File: src/snorri/pathfinding/PathGraph.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
 	private final int width, height;
 
 	@SuppressWarnings("unchecked")
-	public PathGraph(List<Level> levels, int width, int height) {
+	public PathGraph(List<TileLayer> levels, int width, int height) {
```
Comment (tojaroslaw) [inline src/snorri/pathfinding/PathGraph.java]: you should rename these variables to something less confusing than levels

File: src/snorri/world/Editable.java [hunk 2]
```diff
@@ -1,31 +1,37 @@
...
+ *  * Saved in the level editor.
+ *  * Displayed in the level editor.
+ * 	* Edited via various transforms (resizing, reflection, transposition).
+ *
+ */
+public interface Editable extends Loadable, Savable, Renderable {
 	
-	public Level getLevel(int layer);
+	/** Rendering has the same signature as the one in Layer. */
+	public void render(FocusedWindow<?> levelEditor, Graphics2D gr, double deltaTime, boolean b);
```
Comment (tojaroslaw) [inline src/snorri/world/Editable.java]: We should make this extend Renderable and have this method be @Override

File: src/snorri/world/EntityLayer.java [hunk 3]
```diff
@@ -2,113 +2,118 @@
...
+	public boolean canShootOver(Vector position) {
+		return true;
 	}
 
 	@Override
-	public Layer getXReflected() {
-		// TODO Auto-generated method stub
-		return null;
+	public String getFilename() {
+		return "entity.level";
```
Comment (tojaroslaw) [inline src/snorri/world/EntityLayer.java]: this should be entity.layer, entity.level is too confusing

File: src/snorri/world/Layer.java [hunk 4]
```diff
@@ -1,16 +1,30 @@
...
+ * @author tojaroslaw, lambdaviking
+ *
+ */
+public interface Layer extends Renderable {
 
-	public int getHeight();
-	public int getWidth();
+	public boolean canShootOver(Vector position);	
+	
+	// TODO(lambdaviking): Save this stuff.
```
Comment (tojaroslaw) [inline src/snorri/world/Layer.java]: find out what this means

File: src/snorri/world/Layer.java [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
+	static Layer fromYAML(World world, Map<String, Object> params) throws IOException {
+		String type = (String) params.get("type");
+		switch (type) {
+		case "background":
```
Comment (tojaroslaw) [inline src/snorri/world/Layer.java]: maybe we should have an enum or constants, just something to think about, having strings seems bad for code maintainability

File: src/snorri/world/SavableLayer.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * Interface for layers that are Savable.
+ * @author lambdaviking
+ * 
+ */
+public interface SavableLayer extends Layer, Savable {
+	
+	/** Get the filename for saving the Level in the World folder. */
+	public String getFilename();
```
Comment (tojaroslaw) [inline src/snorri/world/SavableLayer.java]: should we have a way to change the filename?

File: src/snorri/world/TileLayer.java [hunk 7]
```diff
@@ -666,4 +645,10 @@ public Tile getOutsideTile() {
 	public void setOutsideTile(Tile outsideTile) {
 		this.outsideTile = outsideTile;
 	}
+
+	@Override
+	public String getFilename() {
+		return "tile.level";
```
Comment (tojaroslaw) [inline src/snorri/world/TileLayer.java]: this should be tile.layer

File: src/snorri/world/World.java [hunk 8]
```diff
@@ -235,14 +177,17 @@ public synchronized boolean delete(Entity e) {
...
 	@Override
 	public void save(File f, boolean recomputeGraphs) throws IOException {
-
+		if (f == null) {
+			throw new IllegalArgumentException("Save path for World must be non-null.");
+		}
+		
 		if (f.exists() && !f.isDirectory()) {
-			throw new IOException("tried to save world to non-directory");
+			throw new IOException("Tried to save world to non-directory.");
```
Comment (tojaroslaw) [inline src/snorri/world/World.java]: grammar????????????????????????????????

File: src/snorri/world/World.java [hunk 9]
```diff
@@ -251,46 +196,61 @@ public void save(File f, boolean recomputeGraphs) throws IOException {
...
+		if (layers == null) {
+			File configFile = new File(f, "config.yml");
+			throw new IllegalArgumentException("No layers specified in " + configFile.getAbsolutePath() + ".");
+		}
+		layers.forEach(params -> {
+			try {
+				Layer layer = Layer.fromYAML(this, params);
+				addLayer(layer);
+			} catch (Exception e) {
+				// FIXME(lambdaviking): Does getting rid of forEach allow us not to use this clause?
```
Comment (tojaroslaw) [inline src/snorri/world/World.java]: what is this for?

File: src/snorri/world/World.java [hunk 10]
```diff
@@ -548,16 +444,41 @@ public Playable getUniverse() {
...
+		return tileLayer;
+	}
+		
+	/** Get a TileLayer by index.
+	 * 
+	 * The generic type T can be used to ensure that the returned layer is of a certain type.
+	 * @param index Index of the layer in the layers array.
+	 * @return The layer whose index in layers is <code>index</code>.
+	 */
+	// TODO(lambdaviking): Figure this out.
```
Comment (tojaroslaw) [inline src/snorri/world/World.java]: What do you need to figure out?

File: src/snorri/pathfinding/PathGraph.java (hunk 1 above)
Comment (viking-sudo-rm) [reply src/snorri/pathfinding/PathGraph.java]: I agree.

File: src/snorri/world/Editable.java (hunk 2 above)
Comment (viking-sudo-rm) [reply src/snorri/world/Editable.java]: Added `@Override` annotation.

File: src/snorri/world/EntityLayer.java (hunk 3 above)
Comment (viking-sudo-rm) [reply src/snorri/world/EntityLayer.java]: Changed. Also changed default value when generating a config.

**Pull request merged**